MC Truth information for MTD

AuroraPerego opened the pull request:
#### PR description:

This PR implements the MC truth information for MTD at the cluster level.
The `simCluster`s creation follows the logic of the `CaloTruthAccumulator` and after that two additional collections are created: the `MtdSimLayerCluster` and the `MtdSimTrackster` (naming adapted from HGCAL). 

- `MtdSimLayerCluster`: these are created with the same logic of the reconstructed clusters. The `simCluster`s are splitted based on the sub detector / layer (for ETL) / module and on the requirement of having hits in adjacents rows / columns. The time associated to these clusters is the average of the hits' time weighted on the energy. These objects will be useful for the Mtd Validation, e.g. to optimize the clustering in the barrel. 
- `MtdSimTrackster`: these are created linking `MtdSimLayerCluster`s associated to the same `simTrack`. The time associated to these tracksters is the time of the first `simHit`. These objects will be useful within the TICL reconstruction to improve linking between MTD and HGCAL objects.

In addition, a new class `MTDGeomUtil` has been created to ease the process of obtaining hits position, layer, module, etc..

This follows the PR #41318 that had the consequence of more `simHit`s being saved in each `simCluster`.

#### PR validation:
The results have been checked on wf 20891.0, looking at single pions.

cc: @fabiocos @felicepantaleo @waredjeb

AuroraPerego committed:
Add MC truth information for MTD clusters

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41662/35537

- This PR adds an extra 152KB to repository

- Found files with invalid states:
  - SimDataFormats/CaloAnalysis/interface/MtdSimClusterFwd.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d
  - RecoLocalFastTime/FTLCommonAlgos/interface/RecHitTools.h:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - RecoLocalFastTime/FTLCommonAlgos/src/RecHitTools.cc:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/interface/MtdSimCluster.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d

cmsbuild commented:
A new Pull Request was created by  @AuroraPerego (Aurora Perego) for master.

It involves the following packages:

- Geometry/MTDGeometryBuilder (**geometry, upgrade**)
- SimCalorimetry/Configuration (**simulation**)
- SimDataFormats/CaloAnalysis (**simulation**)
- SimGeneral/CaloAnalysis (**simulation**)
- SimGeneral/MixingModule (**simulation**)
- Validation/MtdValidation (**dqm**)


@civanch, @Dr15Jones, @bsunanda, @makortel, @nothingface0, @emanueleusai, @mdhildreth, @cmsbuild, @AdrianoDee, @srimanob, @pmandrik, @syuvivida, @tjavaid, @micsucmed, @rvenditti can you please review it and eventually sign? Thanks.
@youyingli, @missirol, @makortel, @felicepantaleo, @rovere, @apsallid, @bsunanda, @thomreis, @wang0jin, @sameasy, @fabiocos, @slomeo this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
-1

**Failed Tests**: RelVals RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32602/summary.html
**COMMIT**: b46895bde0dbaeb77e087cdc7de58ddfff91fa19
**CMSSW**: CMSSW_13_2_X_2023-05-12-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41662/32602/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals

- [23834.999](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32602/runTheMatrix-results/23834.999_TTbar_14TeV+2026D95PU_PMXS1S2PR)
```
----- Begin Fatal Exception 13-May-2023 06:23:52 CEST-----------------------
An exception of category 'StdException' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'PREMIXoutput_step'
   [2] Prefetching for module PoolOutputModule/'PREMIXoutput'
   [3] Calling method for module MixingModule/'mix'
Exception Message:
A std::exception was thrown.
vector::_M_range_check: __n (which is 0) >= this->size() (which is 0)
----- End Fatal Exception -------------------------------------------------
```


## RelVals-INPUT

- [21034.21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32602/runTheMatrixINPUT-results/21034.21_TTbar_14TeV+2026D88PU_ProdLike)```21034.21_TTbar_14TeV+2026D88PU_ProdLike/step2_TTbar_14TeV+2026D88PU_ProdLike.log```
- [21034.114](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32602/runTheMatrixINPUT-results/21034.114_TTbar_14TeV+2026D88PU_OTInefficiency10PC)```21034.114_TTbar_14TeV+2026D88PU_OTInefficiency10PC/step2_TTbar_14TeV+2026D88PU_OTInefficiency10PC.log```
- [21034.999](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32602/runTheMatrixINPUT-results/21034.999_TTbar_14TeV+2026D88PU_PMXS1S2PR)```21034.999_TTbar_14TeV+2026D88PU_PMXS1S2PR/step2_TTbar_14TeV+2026D88PU_PMXS1S2PR.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [21034.99](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32602/runTheMatrixINPUT-results/21034.99_TTbar_14TeV+2026D88PU_PMXS1S2)
- [21034.9921](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32602/runTheMatrixINPUT-results/21034.9921_TTbar_14TeV+2026D88PU_PMXS1S2ProdLike)
- [23834.21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32602/runTheMatrixINPUT-results/23834.21_TTbar_14TeV+2026D95PU_ProdLike)
- [23834.999](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32602/runTheMatrixINPUT-results/23834.999_TTbar_14TeV+2026D95PU_PMXS1S2PR)
- [23834.9921](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32602/runTheMatrixINPUT-results/23834.9921_TTbar_14TeV+2026D95PU_PMXS1S2ProdLike)
- [23834.99](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32602/runTheMatrixINPUT-results/23834.99_TTbar_14TeV+2026D95PU_PMXS1S2)
</details>

rovere commented on SimDataFormats/CaloAnalysis/interface/SimCluster.h:
```diff
@@ -209,11 +223,34 @@ class SimCluster {
+  }
+
+  /** @brief computes the time of the cluster */
+  float computeClusterTime() const {
```
Are we sure such a time assignment makes sense also for calorimeters, like HGCAL? If not and if it's specific to MTD, maybe it would be better to move it to some derived class?

fabiocos replied:
@rovere this calculation is that used in the MTD clusterization, nothing particularly fancy (a energy-weighted time average), but HGCAL experts should comment whether this makes sense of that detector

AuroraPerego replied:
would a new class like `MtdSimCluster` that inherits from the `SimCluster` and adds the time related methods be ok for you?

fabiocos replied:
I believe it is up to main users to reply here, IMHO it might make sense. Alternatively the method could be made subdetector dependent, but I believe you miss this piece of information inside this object, which might be unpractical if you want to extend it beyond the only HGCAL. I let other to comment about this.

fabiocos replied:
@AuroraPerego could you please clarify the strategy? I see that now ```MtdSimLayerCluster``` does not inherit any more from ```SimCluster```, but the time definition is still here. Is it clear if this is ok for all the stakeholders?

rovere commented on SimDataFormats/CaloAnalysis/src/MtdSimCluster.cc:
The naming of this file is confusing. Rename is following the class name it contains.

fabiocos replied:
I agree, no reason to have the file and class names different

rovere commented on SimDataFormats/CaloAnalysis/src/classes_def.xml:
```diff
@@ -8,14 +9,37 @@
-    <version ClassVersion="3" checksum="1773780587"/>
+<class name="std::pair<edm::ProductID,int>" />
+  <class name="SimCluster"  ClassVersion="10">
+   <version ClassVersion="10" checksum="1671921505"/>
```
Do not waste 6 classversion. Just increase it by one.

rovere commented on SimGeneral/CaloAnalysis/plugins/MtdTruthAccumulator.cc:
This class is 90% a copy/paste from the `CaloTruthAccumulator`. Is there any chance we could avoid such a copy/paste?
At the very least, all the definitions, utilities and functions that handle the internal graph representation of the simulated decay could be factored between the 2 approaches.

rovere commented on SimGeneral/MixingModule/python/digitizers_cfi.py:
```diff
@@ -84,7 +84,28 @@
+    maximumSubsequentBunchCrossing = cms.uint32(0),
+
+    simHitCollections = cms.PSet(
+        btl = cms.VInputTag(
```
Is this acronym correct?

fabiocos replied:
since both Barrel and Endcap is handled here, I would say that ```btl``` sounds misleading. @AuroraPerego what about ```mtdCollections```?

fabiocos commented on SimDataFormats/CaloAnalysis/src/MtdSimCluster.cc:
```diff
@@ -0,0 +1,47 @@
+  for (MtdSimLayerCluster::g4t_iterator g4T = tp.g4Track_begin(); g4T != tp.g4Track_end(); ++g4T) {
+    s << " Geant Track Momentum  " << g4T->momentum() << std::endl;
+    s << " Geant Track ID & type " << g4T->trackId() << " " << g4T->type() << std::endl;
+    if (g4T->type() != tp.pdgId()) {
```
May be having everything on a single line makes the printout more friendly to check, too big outputs ends up to be unreadable

AuroraPerego replied:
This is the same format that is used for the `SimCluster` class, but If you prefer it in one line I can change it

fabiocos replied:
@AuroraPerego not a big deal, just my own taste

fabiocos commented on Geometry/MTDGeometryBuilder/src/MTDGeomUtil.cc:
```diff
@@ -0,0 +1,218 @@
+
+using namespace mtd;
+
+namespace {
```
apparently not in this PR...

fabiocos reviewed (commented):
A few comments on top of those of @rovere

fabiocos reviewed (commented):
A few comments on top of those of @rovere

makortel commented on SimDataFormats/CaloAnalysis/src/classes_def.xml:
```diff
@@ -8,14 +9,37 @@
-  <class name="SimCluster"  ClassVersion="4">
-    <version ClassVersion="4" checksum="3868743425"/>
-    <version ClassVersion="3" checksum="1773780587"/>
+<class name="std::pair<edm::ProductID,int>" />
```
This dictionary should be declared in `DataFormats/Provenance`. The package already declares dictionary for `std::pair<edm::ProductID, unsigned int>`, which makes me wonder if it would be feasible to use `unsigned int` here instead of `int`? (I did not look in where and howthis `pair` is used in this PR)

AuroraPerego replied:
In the end I've removed the part of the code that needed this dictionary, so I've removed it as well

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41662/35578

- This PR adds an extra 20KB to repository

- Found files with invalid states:
  - SimDataFormats/CaloAnalysis/interface/MtdSimClusterFwd.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d
  - RecoLocalFastTime/FTLCommonAlgos/interface/RecHitTools.h:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/src/MtdSimCluster.cc:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e
    - Modified: 09c20c3240df802c0a9c4227ebfb512462a5651d, 121983e116f4e92681ac40e233104ae46675d9c7
    - Deleted: 2c0a165069b02653de6025924c3100cea4f6313b
  - RecoLocalFastTime/FTLCommonAlgos/src/RecHitTools.cc:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/interface/MtdSimCluster.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d

cmsbuild commented:
Pull request #41662 was updated. @civanch, @Dr15Jones, @bsunanda, @makortel, @nothingface0, @emanueleusai, @mdhildreth, @cmsbuild, @AdrianoDee, @srimanob, @pmandrik, @syuvivida, @tjavaid, @micsucmed, @rvenditti can you please check and sign again.

emanueleusai commented:
please test

AuroraPerego commented:
> please test

(It will fail again, I'm still working on fixing the error)

cmsbuild commented:
-1

**Failed Tests**: RelVals RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32681/summary.html
**COMMIT**: 2c0a165069b02653de6025924c3100cea4f6313b
**CMSSW**: CMSSW_13_2_X_2023-05-16-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41662/32681/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals

- [23834.999](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32681/runTheMatrix-results/23834.999_TTbar_14TeV+2026D95PU_PMXS1S2PR)```23834.999_TTbar_14TeV+2026D95PU_PMXS1S2PR/step2_TTbar_14TeV+2026D95PU_PMXS1S2PR.log```


## RelVals-INPUT

- [21034.21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32681/runTheMatrixINPUT-results/21034.21_TTbar_14TeV+2026D88PU_ProdLike)```21034.21_TTbar_14TeV+2026D88PU_ProdLike/step2_TTbar_14TeV+2026D88PU_ProdLike.log```
- [21034.114](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32681/runTheMatrixINPUT-results/21034.114_TTbar_14TeV+2026D88PU_OTInefficiency10PC)```21034.114_TTbar_14TeV+2026D88PU_OTInefficiency10PC/step2_TTbar_14TeV+2026D88PU_OTInefficiency10PC.log```
- [21034.999](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32681/runTheMatrixINPUT-results/21034.999_TTbar_14TeV+2026D88PU_PMXS1S2PR)```21034.999_TTbar_14TeV+2026D88PU_PMXS1S2PR/step2_TTbar_14TeV+2026D88PU_PMXS1S2PR.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [21034.99](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32681/runTheMatrixINPUT-results/21034.99_TTbar_14TeV+2026D88PU_PMXS1S2)
- [21034.9921](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32681/runTheMatrixINPUT-results/21034.9921_TTbar_14TeV+2026D88PU_PMXS1S2ProdLike)
- [23834.21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32681/runTheMatrixINPUT-results/23834.21_TTbar_14TeV+2026D95PU_ProdLike)
- [23834.999](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32681/runTheMatrixINPUT-results/23834.999_TTbar_14TeV+2026D95PU_PMXS1S2PR)
- [23834.99](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32681/runTheMatrixINPUT-results/23834.99_TTbar_14TeV+2026D95PU_PMXS1S2)
- [23834.9921](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32681/runTheMatrixINPUT-results/23834.9921_TTbar_14TeV+2026D95PU_PMXS1S2ProdLike)
</details>

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41662/35615

- This PR adds an extra 180KB to repository

- Found files with invalid states:
  - SimDataFormats/CaloAnalysis/interface/MtdSimClusterFwd.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d
  - RecoLocalFastTime/FTLCommonAlgos/interface/RecHitTools.h:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/src/MtdSimCluster.cc:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e
    - Modified: 09c20c3240df802c0a9c4227ebfb512462a5651d, 121983e116f4e92681ac40e233104ae46675d9c7
    - Deleted: 2c0a165069b02653de6025924c3100cea4f6313b
  - RecoLocalFastTime/FTLCommonAlgos/src/RecHitTools.cc:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/interface/MtdSimCluster.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d

cmsbuild commented:
Pull request #41662 was updated. @civanch, @Dr15Jones, @bsunanda, @makortel, @nothingface0, @emanueleusai, @mdhildreth, @AdrianoDee, @srimanob, @pmandrik, @syuvivida, @tjavaid, @micsucmed, @rvenditti can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32751/summary.html
**COMMIT**: d1a44b8b4262bd9f335c740889059418b17b4608
**CMSSW**: CMSSW_13_2_X_2023-05-22-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41662/32751/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 2 lines from the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3221591
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3221563
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-05-22-2300+bcf6f0/57166/triggerResults): no differences found

fabiocos commented on Geometry/MTDGeometryBuilder/src/MTDGeomUtil.cc:
```diff
@@ -0,0 +1,206 @@
+
+// returns the local position as a pair (x, y) - for ETL
+std::pair<float, float> MTDGeomUtil::getPixelInModule(const DetId& id, const int row, const int column) const {
+  ETLDetId detId(id);
```
@AuroraPerego why is only the ETL case considered in this method, while also BTL is treated in the following one?

AuroraPerego replied:
because row and column are needed only for ETL to obtain the pixel position, while for BTL they are retrieved from the topology. So there is no need to use this method for BTL

fabiocos replied:
in principle you are not interested to do it, but the method should be better capable to handle any possible case, or have a protection preventing a BTL id being passed

fabiocos commented on Geometry/MTDGeometryBuilder/src/MTDGeomUtil.cc:
```diff
@@ -0,0 +1,206 @@
+
+float MTDGeomUtil::getPhi(const GlobalPoint& position) const {
+  float phi = atan2(position.y(), position.x());
+  return phi;
```
@AuroraPerego although academic, perhaps better to add a protection as suggested in https://twiki.cern.ch/twiki/bin/view/CMS/CMSConventions#The_azimuthal_angle_phi_in_CMS for the 0,0 case

fabiocos commented on Geometry/MTDGeometryBuilder/src/MTDGeomUtil.cc:
```diff
@@ -0,0 +1,206 @@
+float MTDGeomUtil::getPhi(const DetId& id, const LocalPoint& local_point) const {
+  GlobalPoint position = getPosition(id, local_point);
+  float phi = atan2(position.y(), position.x());
+  return phi;
```
and here as well

fabiocos commented on SimDataFormats/CaloAnalysis/interface/MtdSimLayerCluster.h:
```diff
@@ -0,0 +1,300 @@
+ * @date original date unknown, re-engineering Jan-May 2013
+ */
+class MtdSimLayerCluster {
+  friend std::ostream &operator<<(std::ostream &s, MtdSimLayerCluster const &tp);
```
@AuroraPerego which is the idea behind making this class fully independent on ```SimCluster```, while sharing most of its structure with it?

fabiocos commented on SimGeneral/CaloAnalysis/plugins/MtdTruthAccumulator.cc:
```diff
@@ -0,0 +1,937 @@
+
+    auto update_clu_info = [&](const int &ind) {
+      SimLCenergy += convertUnitsTo(0.001_MeV, m_detIdToTotalSimEnergy[hAndF[ind].first]);
+      SimLCx +=
```
@AuroraPerego this list of energy conversions from GeV to MeV should not be needed here. According to CMS conventions, see https://twiki.cern.ch/twiki/bin/view/CMS/CMSConventions#CMS_units , we should use GeV for energies, and MTD simHits are correctly stored with such a unit for energy. The issue is with MTD rechits, that historically have been erroneously recomputed in MeV. This is the reason of the energy conversions used in the MTD validation code. I would anyway prefer that we limit as much as possible the exception to the convention, which sooner or later I'd like to fix, and move such conversion only where strictly needed by direct comparison with reco hits/clusters (so far validation I would say). The fewer places where changes are done, the easier will be to fix it.

In case please clarify the absolute need for such a conversion in this place.

AuroraPerego replied:
I'll move back to GeV

fabiocos commented on SimDataFormats/CaloAnalysis/interface/MtdSimLayerCluster.h:
```diff
@@ -0,0 +1,300 @@
+
+  /** @brief Gives the total number of SimHits, in the cluster */
+  int numberOfRecHits() const { return hits_.size(); }
+
```
@AuroraPerego this name is misleading, isn't it? I understand there is no reco information in this object, and these two methods above should provide the same result, is it correct?

fabiocos commented on SimDataFormats/CaloAnalysis/interface/MtdSimLayerCluster.h:
```diff
@@ -0,0 +1,300 @@
+
+  /** @brief Returns list of rechit IDs and times for this MtdSimLayerCluster */
+  std::vector<std::pair<uint32_t, float>> hits_and_times() const {
+    assert(hits_.size() == times_.size());
```
this should confirm that hits_ .size() and nsimhits are the same thing

fabiocos commented on SimGeneral/CaloAnalysis/plugins/MtdTruthAccumulator.cc:
```diff
@@ -0,0 +1,937 @@
+      else
+        return geomTools_.getPixelInModule(aa, m_detIdToPos[aa]).second <
+               (int)geomTools_.getPixelInModule(bb, m_detIdToPos[bb]).second;
+    });
```
@AuroraPerego as discussed, this works ok for BTL. In ETL anyway DetId is not enough to identify the hit location, row and column are also needed. And this lexicographic ordering does not necessarily put always close to each other adjacent hits, in case of very high occupancy, do you agree?

AuroraPerego commented:
**Regarding the last commit**
In `MTDGeomUtil.cc` I have:
- added the check on `atan2` not being 0 over 0
- added a throw exception if BTL DetId are passed for an ETL only method
- modified the getPosition to return a pair of local and global position, before the local position was computed to obtain the global position but not returned
- changed the name of the other method that returns only the global position to avoid confusion

In addition, I'm now storing the energy in GeV and not in MeV. 

I have restored `SimCluster.h` and `CaloParticle.h` to their original version and inherit from these -> `MtdSimCluster.h` and `MtdCaloParticle.h`

The common part between `MtdTruthAccumulator.cc` and `CaloTruthAccumulator.cc` has been factored out and put in `DecayGraph.h`

The issue with ETL DetIds not being enough has been addressed by defining an unique number based on DetId, row and column.
```cpp
uint64_t uniqueId = (uint64_t)id.rawId() << 16;
uniqueId |= pixel.first << 8;
uniqueId |= pixel.second;
```
This is used for all the maps in the Accumulator. Then in the `MtdSimCluster.h` there is a method to obtain the original DetId, row and column.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41662/35791

- This PR adds an extra 40KB to repository

- Found files with invalid states:
  - SimDataFormats/CaloAnalysis/interface/MtdSimClusterFwd.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d
  - RecoLocalFastTime/FTLCommonAlgos/interface/RecHitTools.h:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/src/MtdSimCluster.cc:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Modified: 09c20c3240df802c0a9c4227ebfb512462a5651d, 121983e116f4e92681ac40e233104ae46675d9c7
    - Deleted: 2c0a165069b02653de6025924c3100cea4f6313b
  - RecoLocalFastTime/FTLCommonAlgos/src/RecHitTools.cc:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/interface/MtdSimCluster.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41662/35791/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41662/35791/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41662/35792

- This PR adds an extra 40KB to repository

- Found files with invalid states:
  - SimDataFormats/CaloAnalysis/interface/MtdSimClusterFwd.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d
  - RecoLocalFastTime/FTLCommonAlgos/interface/RecHitTools.h:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/src/MtdSimCluster.cc:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Modified: 09c20c3240df802c0a9c4227ebfb512462a5651d, 121983e116f4e92681ac40e233104ae46675d9c7
    - Deleted: 2c0a165069b02653de6025924c3100cea4f6313b
  - RecoLocalFastTime/FTLCommonAlgos/src/RecHitTools.cc:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/interface/MtdSimCluster.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d
    - Modified: da317b54dd1d286de2eb29e793b6129b744df906

cmsbuild commented:
Pull request #41662 was updated. @civanch, @Dr15Jones, @bsunanda, @makortel, @nothingface0, @emanueleusai, @mdhildreth, @cmsbuild, @AdrianoDee, @srimanob, @pmandrik, @syuvivida, @tjavaid, @micsucmed, @rvenditti can you please check and sign again.

perrotta commented:
@cmsbuild please test

cmsbuild commented:
-1

**Failed Tests**:   HeaderConsistency 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/32998/summary.html
**COMMIT**: da317b54dd1d286de2eb29e793b6129b744df906
**CMSSW**: CMSSW_13_2_X_2023-06-05-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41662/32998/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 23 lines to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3219909
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3219884
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-06-05-1100+bcf6f0/57361/triggerResults): no differences found

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41662/35808

- This PR adds an extra 28KB to repository

- Found files with invalid states:
  - SimDataFormats/CaloAnalysis/interface/MtdSimClusterFwd.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d
  - RecoLocalFastTime/FTLCommonAlgos/interface/RecHitTools.h:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/src/MtdSimCluster.cc:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Modified: 09c20c3240df802c0a9c4227ebfb512462a5651d, 121983e116f4e92681ac40e233104ae46675d9c7
    - Deleted: 2c0a165069b02653de6025924c3100cea4f6313b
  - RecoLocalFastTime/FTLCommonAlgos/src/RecHitTools.cc:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/interface/MtdSimCluster.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d
    - Modified: da317b54dd1d286de2eb29e793b6129b744df906

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41662/35828

- This PR adds an extra 220KB to repository

- Found files with invalid states:
  - SimDataFormats/CaloAnalysis/interface/MtdSimClusterFwd.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d
  - RecoLocalFastTime/FTLCommonAlgos/interface/RecHitTools.h:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/src/MtdSimCluster.cc:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Modified: 09c20c3240df802c0a9c4227ebfb512462a5651d, 121983e116f4e92681ac40e233104ae46675d9c7
    - Deleted: 2c0a165069b02653de6025924c3100cea4f6313b
  - RecoLocalFastTime/FTLCommonAlgos/src/RecHitTools.cc:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/interface/MtdSimCluster.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d
    - Modified: dabe885b0a4b31affc361254bfad786b5395502d

cmsbuild commented:
Pull request #41662 was updated. @civanch, @Dr15Jones, @bsunanda, @makortel, @mdhildreth, @cmsbuild, @AdrianoDee, @srimanob can you please check and sign again.

fabiocos commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/33032/summary.html
**COMMIT**: dabe885b0a4b31affc361254bfad786b5395502d
**CMSSW**: CMSSW_13_2_X_2023-06-07-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41662/33032/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 15 lines from the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 49 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3219909
 - DQMHistoTests: Total failures: 2225
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3217662
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-06-07-1100+bcf6f0/57385/triggerResults): no differences found

civanch commented:
+1

differences are in MkFit WF

srimanob commented:
+Upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
@AuroraPerego thanks for addressing the various comments. I believe the good is in a good enough shape to be integrated and tested more extensively in practical use cases. Next step, from the MTD point of view, will be the creation of association maps between ```MtdSimLayerCluster``` and ```TrackingParticle``` collections, that can be then used for instance for validation studies.

fabiocos commented:
@AuroraPerego since you are expanding the event size, did you check the fractional added size taken by these new collections in your benchmark events?

AuroraPerego commented:
> @AuroraPerego since you are expanding the event size, did you check the fractional added size taken by these new collections in your benchmark events?

This is what I see by doing `edmEventSize` (wf 23834.999):
| Branch Name | Average Uncompressed Size (Bytes/Event) | Average Compressed Size (Bytes/Event) |
|-------|--------|-------|
| MtdCaloParticles_mix_MergedMtdTruth_HLT. | 65434.8 | 16700.9 |
| MtdSimClusters_mix_MergedMtdTruth_HLT. | 128591 | 46090.3 |
| MtdSimLayerClusters_mix_MergedMtdTruthLC_HLT. | 214323 | 52472.6 |
| MtdSimTracksters_mix_MergedMtdTruthST_HLT. | 130851 | 42782.3 |
| ulongfloatstdpairs_mix_MergedMtdTruth_HLT. | 22928.2 | 14583.2 |

That should correspond to a fraction of 0.53% in the Uncompressed Size and 0.91% in the Compressed Size

perrotta commented on SimDataFormats/CaloAnalysis/interface/MtdSimCluster.h:
```diff
@@ -0,0 +1,111 @@
+
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
+#include "DataFormats/Math/interface/LorentzVector.h"
+#include "DataFormats/Math/interface/Point3D.h"
+#include "DataFormats/Math/interface/Vector3D.h"
```
```suggestion
```
are they needed?

perrotta commented on SimDataFormats/CaloAnalysis/interface/MtdSimLayerCluster.h:
```diff
@@ -0,0 +1,70 @@
+
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
+#include "DataFormats/Math/interface/LorentzVector.h"
+#include "DataFormats/Math/interface/Point3D.h"
+#include "DataFormats/Math/interface/Vector3D.h"
```
```suggestion
```
are they needed?

perrotta commented on SimDataFormats/CaloAnalysis/interface/MtdSimLayerCluster.h:
```diff
@@ -0,0 +1,70 @@
+#ifndef SimDataFormats_CaloAnalysis_MtdSimLayerCluster_h
+#define SimDataFormats_CaloAnalysis_MtdSimLayerCluster_h
+
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
```
```suggestion
```
in the implementation file?

perrotta commented on SimDataFormats/CaloAnalysis/interface/MtdSimTrackster.h:
```diff
@@ -0,0 +1,51 @@
+#define SimDataFormats_MtdMtdSimTrackster_h
+
+#include "DataFormats/GeometryVector/interface/GlobalPoint.h"
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
```
```suggestion
```
in the implementation file?

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41662/35896

- This PR adds an extra 32KB to repository

- Found files with invalid states:
  - SimDataFormats/CaloAnalysis/interface/MtdSimClusterFwd.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d
  - RecoLocalFastTime/FTLCommonAlgos/interface/RecHitTools.h:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/src/MtdSimCluster.cc:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Modified: 09c20c3240df802c0a9c4227ebfb512462a5651d, 121983e116f4e92681ac40e233104ae46675d9c7, f5bdbc75dc011ab8d38dab11b623e1b28e46aafc
    - Deleted: 2c0a165069b02653de6025924c3100cea4f6313b
  - RecoLocalFastTime/FTLCommonAlgos/src/RecHitTools.cc:
    - Added: 5e71e76710bd6893fb182b1f49410fd50d8dd11a
    - Deleted: f39c14b6fcb0b447c31b45bbfabfa922ba2d1aca
  - SimDataFormats/CaloAnalysis/interface/MtdSimCluster.h:
    - Added: b232a3d7eb57622220303eceb86a3fa7a002fc1e, 7efb6f6c8392ece62ff421dfb6e9e181889f98df
    - Deleted: 09c20c3240df802c0a9c4227ebfb512462a5651d
    - Modified: dabe885b0a4b31affc361254bfad786b5395502d, f5bdbc75dc011ab8d38dab11b623e1b28e46aafc

cmsbuild commented:
Pull request #41662 was updated. @civanch, @Dr15Jones, @bsunanda, @makortel, @mdhildreth, @cmsbuild, @AdrianoDee, @srimanob can you please check and sign again.

civanch commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bcf6f0/33110/summary.html
**COMMIT**: f5bdbc75dc011ab8d38dab11b623e1b28e46aafc
**CMSSW**: CMSSW_13_2_X_2023-06-12-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41662/33110/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 1 lines from the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 55 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3190910
 - DQMHistoTests: Total failures: 2231
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3188657
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-06-12-2300+bcf6f0/57453/triggerResults): no differences found

fabiocos commented on Geometry/MTDGeometryBuilder/src/MTDGeomUtil.cc:
```diff
@@ -0,0 +1,209 @@
+                                               << detId.rawId() << ") is invalid!" << std::dec << std::endl;
+    global_point = thedet->toGlobal(local_point);
+  } else {
+    throw cms::Exception("mtd::MTDGeomUtil") << "detId " << id.rawId() << " not in MTD" << std::endl;
```
same argument applies here

AuroraPerego commented on Geometry/MTDGeometryBuilder/src/MTDGeomUtil.cc:
```diff
@@ -0,0 +1,209 @@
+    local_point = LocalPoint(topo.localX(row), topo.localY(column), 0.);
+    global_point = thedet->toGlobal(local_point);
+  } else {
+    throw cms::Exception("mtd::MTDGeomUtil") << "detId " << id.rawId() << " not in MTD" << std::endl;
```
I've removed the `else` that cannot be reached (here and below). `isETL` contains a check on the `DetId` being in MTD, so that should be enough to handle any situation.

perrotta commented:
@AuroraPerego since you did not get the signatures since your latest commit, it is probably worth squashing all commits here into one or a few, and get rid as such of the files with invalid state in github. I imagine @fabiocos can help you with it, if needed.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41662/35951

- This PR adds an extra 12KB to repository

cmsbuild commented:
Pull request #41662 was updated. @civanch, @Dr15Jones, @bsunanda, @makortel, @mdhildreth, @cmsbuild, @AdrianoDee, @srimanob can you please check and sign again.

srimanob commented:
@cmsbuild please test

srimanob commented:
+Upgrade

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
